Blog post for form-associated custom elements. (#1287)
GoogleChrome/web.dev: src/site/content/en/blog/more-capable-form-controls/index.md

---

> title: "More capable form controls"

Reviewer: Do you think it's worth highlighting that this relates to custom elements in the title?

Reviewer: ah I see, it's actually talking about two things: `formdata` and custom elements. Maybe 'More capable form controls' is ok then. But it would still be nice to beef up the subhead if possible.

Author: Well, only one of the two APIs is limited to custom elements. That API takes up the bulk of the article, it's true... Perhaps I should add that detail in the subhead, instead?

> subhead: New APIs for form controls

Reviewer: The subhead appears below the headline and is a chance to add some additional perspective to the article. Similar to this medium post:

![image](https://user-images.githubusercontent.com/1066253/62667787-1270bd00-b93e-11e9-92e0-2a6d083572b3.png)

Do you think there's anything you might want to say there to help folks understand how this change improves the usage of custom elements?

*   The Form-associated custom elements API lets custom elements act more like built-in form controls.

These two features can be used to create new kinds of controls that work better with native web forms.

{% Aside %}Building custom form controls is an advanced topic. This article assumes a certain knowledge of forms and form controls. When building a custom form control, there are many factors to consider, especially making sure that your controls are accessible to all users. To learn more about forms, go to the [MDN guide on forms](https://developer.mozilla.org/en-US/docs/Learn/HTML/Forms).{% endAside %} 

## Event-based API

The `formdata` event is a low-level API that lets any JavaScript code participate in form submission. The mechanism works like this:

1.  You add a `formdata` event listener to the form you want to interact with.
1.  When a user clicks the submit button, the form fires a `formdata` event, which includes a <code>[FormData](https://developer.mozilla.org/en-US/docs/Web/API/FormData)</code> object that holds all of the data being submitted.
1.  Each <code>formdata</code> listener gets a chance to add to or modify the data before the form is submitted. 

> Here's an example of sending a single value in an <code>formdata</code> event listener:

Reviewer: ```suggestion
Here's an example of sending a single value in an `formdata` event listener:
```

```js
const form = document.querySelector('form');
// FormData event is sent on <form> submission, before transmission.
// The event has a formData property
form.addEventListener('formdata', ({formData}) => {
  // https://developer.mozilla.org/en-US/docs/Web/API/FormData
  formData.append('my-input', myInputValue);
});
```

> Try this out using our example on Glitch. Be sure to run it on Chrome 77+ (Canary or Beta) to see the API in action.

Reviewer: nice! 🎉

<div class="glitch-embed-wrap" style="height: 420px; width: 100%;">
  <iframe
    allow="geolocation; microphone; camera; midi; encrypted-media"
    src="https://glitch.com/embed/#!/embed/formdata-event?path=index.html&amp;previewSize=0"
    alt="formdata event demo on Glitch"
    style="height: 100%; width: 100%; border: 0;">
  </iframe>
</div>

## Form-associated custom elements

The event-based API is simple, but it only allows you to interact with the submission process.

Native form controls participate in many parts of the form lifecycle besides submission. Form-aware custom elements aim to bridge the gap between custom widgets and native controls. Form-aware custom elements match many of the features of native form elements:

*   When you place a form-associated custom element inside a `<form>`, it's automatically associated with the form, like a native control.

> *   The element can be labeled using a  `<label>` element.

Reviewer: ```suggestion
*   The element can be labeled using a `<label>` element.
```

### Defining a form-associated custom element

Defining a form-associated custom element requires two additions to your custom element:

*   Add a static `formAssociated` property to your custom element class. This tells the browser to treat the element like a form control.
*   Call the `attachInternals()` method on the element to get access to extra methods and properties for form controls, like `setFormValue` and `setValidity`.

Here's how those two items fit into a basic custom element definition: \

```js
// Form-associated custom elements must be autonomous custom elements--
// meaning they must extend HTMLElement, not one of its subclasses.
class MyCounter extends HTMLElement {

  // Identify the element as a form-associated custom element
  static formAssociated = true;

  constructor() {
    super();
    // Get access to the internal form control APIs
    this._internals = this.attachInternals();
    // internal value for this control
    this._value = 0;
  }

>   // Form controls usually expose a "value"  property

Reviewer: ```suggestion
  // Form controls usually expose a "value" property
```

  get value() { return this._value; }
  set value(v) { this._value = v; }

  // The following properties and methods aren't strictly required, 
  // but native form controls provide them. Providing them helps
  // ensure consistency with native controls.
  get form() { return this._internals.form; }
  get name() { return this.getAttribute('name'); }
  get type() { return this.localName; }
  get validity() {return this.internals_.validity; }
  get validationMessage() {return this.internals_.validationMessage; }
  get willValidate() {return this.internals_.willValidate; }

  checkValidity() { return this.internals_.checkValidity(); }
  reportValidity() {return this.internals_.reportValidity(); }

>   ... 

Reviewer: ```suggestion
  … 
```

}

> customElements.register('my-counter', MyCounter);

Reviewer: ```suggestion
customElements.define('my-counter', MyCounter);
```

Author: ZOMG. How'd I get that in there? 😂

*   A string value.
*   A [`File`](https://developer.mozilla.org/en-US/docs/Web/API/File) object.
*   A [`FormData`](https://developer.mozilla.org/en-US/docs/Web/API/FormData) object. You can use a `FormData` object to pass multiple values (for example, a credit card input control might pass a card number, expiration date, and verification code).

To set a simple value:

```js
this._internals.setFormValue(this._value);
```

To set multiple values, you can do something like this:

```js
// Use the control's name as the base name for submitted data
const n = this.getAttribute('name');
const entries = new FormData();
entries.append(n + '-first-name', this.firstName_);
entries.append(n + '-last-name', this.lastName_);
this.internals_.setFormValue(entries);
```

{% Aside %}The `setFormValue` method takes a second, optional `state` parameter, used to store the internal state of the control. For more information, see [Restoring form state](#restoring-form-state).{% endAside %}

### Input validation

Your control can also participate in form validation by calling the `setValidity`
method on the internals object.

```js
// Assume this is called whenever the internal value is updated

> onUpdateValue() {

Reviewer: The indentation in this example seems a bit off?

![image](https://user-images.githubusercontent.com/1066253/62668480-e60a7000-b940-11e9-893f-f31b76af22e6.png)

Author: Yep. Last `}` should be in column 1. The wrapped conditional doesn't help any, but I think it looks better this way than if we force people to scroll.

>   // 

Reviewer: ```suggestion

```
should this be removed?

Author: Yeah... Was going to add a comment here, but never got it right, alas.

  if (!this.matches(':disabled') && this.hasAttribute('required') &&
      this._value < 0) {
    this._internals.setValidity({customError: true}, 'Value cannot be negative.');
  } 
  else {
    this._internals.setValidity({});
  }
  this.internals.setFormValue(this._value);
  }
```

You can style a form-associated custom element with the `:valid` and `:invalid`
pseudoclasses, just like a built-in form control.

> ### Lifecycle callbacks

Reviewer: I was curious what you think about formatting the section like this:

```
### Lifecycle callbacks

A form-associated custom element API includes a set of extra lifecycle callbacks to tie in to the form lifecycle. The callbacks are optional: only implement a callback if your element needs to do something at that point in the lifecycle.

#### `formAssociatedCallback(form)`

Called when the browser associates the element with a form element, or disassociates the element from a form element.

#### `formDisabledCallback(disabled)`

Called after the `disabled` state of the element changes, either because the `disabled` attribute of this element was added or removed; or because the `disabled` state changed on a `<fieldset>` that's an ancestor of this element. The `disabled` parameter represents the new [disabled state](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#concept-fe-disabled) of the element. The element may, for example, disable elements in its shadow DOM when it is disabled.

#### `formResetCallback()`

Called after the form is reset. The element should reset itself to some kind of default state. For native inputs, this usually involves setting the `value` property to match the `value` attribute set in markup (or in the case of a checkbox, setting the `checked` property to match the `checked` attribute.

#### `formStateRestoreCallback(state, mode)`

Called in one of two circumstances:

* When the browser restores the state of the element (for example, after a navigation, or when the browser restarts).  The <code>mode</code> argument is <code>"restore"</code> in this case.
* When the browser's input-assist features such as form autofilling sets a value. The <code>mode</code> argument is <code>"autocomplete"</code> in this case.

The type of the first argument depends on how the <code>setFormValue</code> method was called. For more details, see [Restoring form state](#restoring-form-state).
```

It'll look something like this:

![image](https://user-images.githubusercontent.com/1066253/62668841-3fbf6a00-b942-11e9-8392-4a4fce50b872.png)

Author: Hm... Yeah, I could do that. I had a variation of that, but the headers kind of duplicated the callback names. This could work. 

It's a little weird because there's another H4 after the callbacks called "Restoring form state." It's the gnarly details of how formStateRestoreCallback interacts with setFormValue. It felt like too much information to cram into the callback section. I'll take another look at it.

Author: Done.

    The type of the first argument depends on how the <code>setFormValue</code> method was called. For more details, see [Restoring form state](#restoring-form-state).

#### Restoring form state {: #restoring-form-state }

Under some circumstances—like when navigating back to a page, or restarting the browser, the brower may try to restore the state of a form to the state the user left it in.

For a form-associated custom element, the restored state comes from the value(s) you pass to the `setFormValue()` method. You can call the method with a single value parameter, as shown in the [earlier examples](#set-a-value), or with two parameters:

```js
this._internals.setFormValue(value, state);
```

The `value` represents the submittable value of the control. The optional `state` parameter is an _internal_ representation of the state of the control, which can include data that doesn't get sent to the server. The `state` parameter takes the same types as the `value` parameter—it can be a string, `File`, or `FormData` object.

> The `state` parameter is useful when you can't restore a control's state based on the value alone. For example, suppose you create a color picker with multiple modes: a  palette or an RGB color wheel. The submittable _value_ would  be the selected color in a canonical form, like '#7fff00'. But to restore the control to a specific state, you'd also need to know which mode it was in, so the _state_ might look like 'palette/#7fff00'.  

Reviewer: ```suggestion
The `state` parameter is useful when you can't restore a control's state based on the value alone. For example, suppose you create a color picker with multiple modes: a  palette or an RGB color wheel. The submittable _value_ would  be the selected color in a canonical form, like `"#7fff00"`. But to restore the control to a specific state, you'd also need to know which mode it was in, so the _state_ might look like `"palette/#7fff00"`.  
```

```

Your code would need to restore its state based on the stored state value.

```js
formStateRestoreCallback(state, mode) {
  if (mode == 'restore') {
    // expects a state parameter in the form 'controlMode/value'
    [controlMode, value] = state.split('/');
    this._mode = controlMode;
    this._value = value;
  }
  // Chrome currently doesn't handle autofill for form-associated 
  // custom elements. In the autofill case, you might need to handle
  // a raw value.
}
```

In the case of a simpler control (for example a number input), the value is probably sufficient to restore the control to its previous state. If you omit `state` when calling `setFormValue`, then the value is passed to `formStateRestoreCallback`. 

```js
formStateRestoreCallback(state, mode) {
  // Simple case, restore the saved value
  this._value = state;
}
```

### A working example

The following example puts together many of the features of form-associated custom elements.
Be sure to run it on Chrome 77+ (Canary or Beta) to see the API in action.

<div class="glitch-embed-wrap" style="height: 420px; width: 100%;">
  <iframe
    allow="geolocation; microphone; camera; midi; encrypted-media"

>     src="https://glitch.com/embed/#!/embed/form-associated-ce?path=my-control.js&amp;previewSize=0"

Reviewer: ```suggestion
    src="https://glitch.com/embed/#!/embed/form-associated-ce?path=public%2Fmy-control.js&amp;previewSize=0"
```

    style="height: 100%; width: 100%; border: 0;">
  </iframe>
</div>

## Feature detection 

You can use feature detection to determine whether the `formdata` event and form-associated custom elements are available. There are currently no polyfills released for either feature. In both cases, you can fall back to adding a hidden form element to propagate the control's value to the form. Many of the more advanced features of form-associated custom elements will likely be difficult or impossible to polyfill.

```js
if ('FormDataEvent' in window) {
  // formdata event is supported
}

if ('ElementInternals' in window && 
    'setFormData' in window.ElementInternals) {
  // Form-associated custom elements are supported
}
```

## Conclusion

The `formdata` event and form-associated custom elements provide new tools for creating custom form controls. 

The `formdata` event doesn't give you any new capabilities, but it gives you an interface for adding your form data to the submit process, without having to create a hidden `<input>` element.

The form-associated custom elements API provides a new set of capabilities for making custom form controls that work like built-in form controls.

> _Hero image by Oudom Pravat on Unsplash._

Reviewer: Just curious if you're able to link this back to the original image? In the future I may use these when we 2x all the headers.

Author: Sure. I hadn't seen anyone else doing that, so I didn't include a link... (In fact, I had to search long and hard to find anyone crediting their hero image.)

For the sake of consistency, how do you want to link it up?

_Hero image by Oudom Pravat on [Unsplash](https://unsplash.com/photos/yQi4mAFmRew)._

Or

_[Hero image](https://unsplash.com/photos/yQi4mAFmRew) by Oudom Pravat on Unsplash._

Or 

[_Hero image by Oudom Pravat on Unsplash._](https://unsplash.com/photos/yQi4mAFmRew)

Author: I went with #1 here, can update if desired.

---
title: "More capable form controls"

> subhead: Build custom elements that work like native form controls 

Reviewer: I guess the problem now is it doesn't make it clear that formdata is also mentioned 😅

What about something like... "With new events, and custom elements APIs, participating in forms just got a lot easier."

Author: 👍 👍 
Tried to work both in w/o luck. You win!

authors:
  - arthurevans
date: 2019-08-08
hero: hero.jpg
# You can adjust the position of your hero image with this property.
# Values: top | bottom | center (default)
# hero_position: bottom
alt: DJ mixer controls.
description: |
  This post describes some new web platform features that make it easier to build 

>   custom elements that work like native form controls. 

Author: Check.

tags:
  - post # post is a required tag for the article to show up in the blog.
  - forms
  - web-components
---

Many developers build custom form controls, either to provide controls that aren't built in to the browser, or to customize the look and feel beyond what's possible with the native form controls. 

> However, if can be difficult to replicate the features of built-in HTML form controls. For example, when you add an `<input>` element to a form:

Author: No, they're not requirements. They're examples of things the browser does automatically when you add an `<input>` element. They're features that built-in form controls get for free, and custom form controls don't.

> However, if can be difficult to replicate the features of built-in HTML form controls. For example, when you add an `<input>` element to a form:

Reviewer 2: I would suggest something like the phrase 'among other things'.

> However, if can be difficult to replicate the features of built-in HTML form controls. For example, when you add an `<input>` element to a form:

Author: I changed this to 

For example, when you add an `<input>` element to a form, several things happen automatically:

But I think we could do better. Maybe. 

However, if can be difficult to replicate the features of built-in HTML form controls. Here are 
some of the things that happen automatically when you add an `<input>` element to a form:

> However, if can be difficult to replicate the features of built-in HTML form controls. For example, when you add an `<input>` element to a form:

Author: Or, "For example, some of the things that happen ..." Or just "Among the things that happen ..."

> However, if can be difficult to replicate the features of built-in HTML form controls. For example, when you add an `<input>` element to a form:

Reviewer 2: Either one is good.

> However, if can be difficult to replicate the features of built-in HTML form controls. For example, when you add an `<input>` element to a form:

Author: Better:

Consider some of the features an `<input>` element gets automatically when you add it to a form:

> However, if can be difficult to replicate the features of built-in HTML form controls. For example, when you add an `<input>` element to a form:

Reviewer 2: It just jumped from good to Awesome.

> However, if can be difficult to replicate the features of built-in HTML form controls. For example, when you add an `<input>` element to a form:

Author: Updated this. Also tweaked the following wording a little.

## Event-based API

The `formdata` event is a low-level API that lets any JavaScript code participate in form submission. The mechanism works like this:

1.  You add a `formdata` event listener to the form you want to interact with.
1.  When a user clicks the submit button, the form fires a `formdata` event, which includes a [`FormData`](https://developer.mozilla.org/en-US/docs/Web/API/FormData) object that holds all of the data being submitted.
1.  Each `formdata` listener gets a chance to add to or modify the data before the form is submitted. 

Here's an example of sending a single value in a `formdata` event listener:

```js
const form = document.querySelector('form');
// FormData event is sent on <form> submission, before transmission.
// The event has a formData property
form.addEventListener('formdata', ({formData}) => {
  // https://developer.mozilla.org/en-US/docs/Web/API/FormData
  formData.append('my-input', myInputValue);
});
```

> Try this out using our example on Glitch. Be sure to run it on Chrome 77+ (Canary or Beta) to see the API in action.

Author: SG.

*   The element can set a value that's automatically submitted with the form.
*   The element can set a flag indicating whether or not it has valid input. If one of the form controls has invalid input, the form can't be submitted.
*   The element can provide callbacks for various parts of the form lifecycle—such as when the form is disabled or reset to its default state.
*   The element supports the standard CSS pseudoclasses for form controls, like `:disabled` and `:invalid`.

That's a lot of features! This article won't touch on all of them, but will describe the basics needed to integrate your custom element with a form.

{% Aside %}This section assumes a basic familiarity with custom elements. For an introduction to custom elements, see [Custom Elements v1: Reusable Web Components](https://developers.google.com/web/fundamentals/web-components/customelements) on Web Fundamentals.{% endAside %}

### Defining a form-associated custom element

Defining a form-associated custom element requires two additions to your custom element:

*   Add a static `formAssociated` property to your custom element class. This tells the browser to treat the element like a form control.

> *   Call the `attachInternals()` method on the element to get access to extra methods and properties for form controls, like `setFormValue()` and `setValidity()`.

Author: Good call. This is really about the scaffolding or boilerplate for the element, so let me take another look at this list.

#### `void formResetCallback()`

Called after the form is reset. The element should reset itself to some kind of default state. For native inputs, this usually involves setting the `value` property to match the `value` attribute set in markup (or in the case of a checkbox, setting the `checked` property to match the `checked` attribute.

#### `void formStateRestoreCallback(state, mode)`

Called in one of two circumstances:

*   When the browser restores the state of the element (for example, after a navigation, or when the browser restarts).  The <code>mode</code> argument is <code>"restore"</code> in this case.
*   When the browser's input-assist features such as form autofilling sets a value. The <code>mode</code> argument is <code>"autocomplete"</code> in this case.

The type of the first argument depends on how the `setFormValue()` method was called. For more details, see [Restoring form state](#restoring-form-state).

### Restoring form state {: #restoring-form-state }

> Under some circumstances—like when navigating back to a page, or restarting the browser, the brower may try to restore the state of a form to the state the user left it in.

Reviewer 2: 'brower' -> 'browser'

'restore the state of a form to the state...' -> 'restore the form to the state...'

Author: 👍

For a form-associated custom element, the restored state comes from the value(s) you pass to the `setFormValue()` method. You can call the method with a single value parameter, as shown in the [earlier examples](#set-a-value), or with two parameters:

```js
this._internals.setFormValue(value, state);
```

The `value` represents the submittable value of the control. The optional `state` parameter is an _internal_ representation of the state of the control, which can include data that doesn't get sent to the server. The `state` parameter takes the same types as the `value` parameter—it can be a string, `File`, or `FormData` object.

> The `state` parameter is useful when you can't restore a control's state based on the value alone. For example, suppose you create a color picker with multiple modes: a  palette or an RGB color wheel. The submittable _value_ would  be the selected color in a canonical form, like `"#7fff00"`. But to restore the control to a specific state, you'd also need to know which mode it was in, so the _state_ might look like `"palette/#7fff00"`.  

Reviewer 2: Delete the extra space toward the beginning of the third line.

```js
this._internals.setFormValue(this._value, 
    this._mode + '/' + this._value);
```

Your code would need to restore its state based on the stored state value.

```js
formStateRestoreCallback(state, mode) {
  if (mode == 'restore') {
    // expects a state parameter in the form 'controlMode/value'
    [controlMode, value] = state.split('/');
    this._mode = controlMode;
    this._value = value;
  }
  // Chrome currently doesn't handle autofill for form-associated 
  // custom elements. In the autofill case, you might need to handle
  // a raw value.
}
```

> In the case of a simpler control (for example a number input), the value is probably sufficient to restore the control to its previous state. If you omit `state` when calling `setFormValue`, then the value is passed to `formStateRestoreCallback`. 

Reviewer 2: Please add parens to the function names.

Author: Thanks. I think I've got all of these now.
